api/v1/standalone-dns-proxy/standalone-dns-proxy.proto

```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+syntax = "proto3";
+
+package standalonednsproxy;
+
+option go_package = "github.com/cilium/cilium/api/v1/standalonednsproxy";
+
+service FQDNData {
+  rpc SubscribeToDNSPolicies(stream  DNSPoliciesResult) returns (stream DNSPolicies) {}
```

Comment:
A suggestion that occurred to me while reading this proto:
Let's formulate this RPC a bit more generally. AFAIU, the SDS needs to enforce egress DNS policy. To do so, it needs access to policy state, which it streams from the agent. Since all it sees are source IPs of outgoing DNS queries, it needs to associate that query with an endpoint. The CFP suggested doing so via the ipcache BPF map, but concerns were voiced regarding making a BPF map a de-facto public API, which I think we can elegantly avoid by streaming more policy state from the agent. Concretely, we can make this a `StreamPolicyState(stream PolicyStateAck) returns (stream PolicyState)` (naming TBD) and make `PolicyState` be a ~union of different objects (endpoints/policies/...) needed to enforce the egress policy. Said differently, instead of looking at the ipcache for association and getting "just" policies via this API, we push all necessary policy information from the agent into the SDS via this API and don't interact with BPF at all. This gives us similar capabilities while the agent is down, and allows for extensions if ever we're interested in making the SDS more capable during agent down-time.

Reply:
I understand how the BPF map lookup seems simpler, but only since you have not addressed the difficult question: How does the SDP handle agent upgrades, which may change the ipcache map format?

The problem I'm thinking of is:

1. The agent and SDP are running.
2. The cluster operators want to perform an upgrade of cilium. Now, thanks to the SDP, they expect DNS traffic and toFQDN policies to continue working. To me, this implies that the SDP and the cilium-agent have to have a different life-cycle; we cannot upgrade both at the same time.
3. The agent upgrade changes the format of the ipcache bpf map
4. Since the SDP image remains the same, it will not be able to correctly access the ipcache map?


Given the above and that we're _already_ streaming policy state towards the SDP, it seems easier to add more state to the SDP than handling the upgrade problem. In the end, it seems like all that's needed is to subscribe to receive ipcache and EP updates.

Regarding connectivity: I would think that the SDP and cilium-agent are running on the same node, and hence have fast and reliable connectivity? It seems to me that failure cases where the SDP doesn't consume the stream or where the cilium-agent crashes need to be handled no matter what we actually send over gRPC.

Reply:
> How does the SDP handle agent upgrades, which may change the ipcache map format?

Regarding upgrade, SDP has to be upgraded first before agent and new SDP should handle both old and new bpf map format. Based on some heuristics (eg: key/value size), new SDP should detect if its old or new format and use appropriately (last section of CFP covers this)

>  It seems to me that failure cases where the SDP doesn't consume the stream or where the cilium-agent crashes need to be handled no matter what we actually send over gRPC.

In that case, SDP has to maintain local cache of ipcache mappings so that when agent is down, it has to enforce L7 dns policy to allow dns request to be forwarded or not. In case of reading directly from bpf map, SDP doesn't need to maintain any cache and always read from bpf map directly even if agent is down.

Reply:
(discussed out of band, will discuss further in the sig-policy meeting tomorrow)

Reply:
(Couldn't get time to talk in sig-policy sync.)

If we rely on GRPC api to send ip-identity updates, then sending all ipcache updates on any remote endpoint label/ip change would be inefficient right?. It would be better if agent sends ipcache snapshot only when fqdn related ip/identity changes. FQDN policy requires src identity and also remote destination identity (dnsserver) to enforce L7 dns policy. So it need to be aware only if local src endpoint or remote dnsserver ip/label change. The complexity is computing this map of identities to be tracked. So, agent needs to have a map to track local endpoint's identity which enforcing FQDN policy and then all of its egress destination identity (since we can't deterministically decide dnsserver) and if any of this identity changes, only then send ip-identity snapshot to SDP. So on every policy update or ipcache updates, agent need to compute if it affects the identity in that set, update the set and then send snapshot to SDP. Need to explore cilium codebase to see if this is feasible.

Reply:
In the end I think it's a question of what you think happens more often: DNS lookups or identity updates. If there's significantly more DNS traffic than there is identity changes, subscribing to the relevant information via the agent is cheaper, since you can do "ipcache" lookups in userspace and hence avoid doing a syscall on every dns message. In addition, the benefit of this solution is that you don't have to worry about BPF map format changes.
On the other hand, if you believe that the identity churn is large enough that the overhead would be too large and we'd have to start computing the relevant set etc, then looking into the ipcache is likely simpler.

> Regarding upgrade, SDP has to be upgraded first before agent and new SDP should handle both old and new bpf map format. Based on some heuristics (eg: key/value size), new SDP should detect if its old or new format and use appropriately (last section of CFP covers this)

I think this is worth spelling out more. It's not clear to me yet.

> SDP has to be upgraded first

Upgraded to where? Does it need to be upgraded to the last patch version of the current minor, or to the minor you're wanting to upgrade cilium to? Or said differently, what do you envision the version compatibility of the SDP related to cilium to be? (This is assuming the SDP is versioned with cilium?)

> first

If you don't know of this requirement, what happens? Is there a way we can detect that the user is doing it in the wrong order, and prevent it?

Also, IIRC, the agent recreates the ipcache map on startup and then atomically swaps it over. How is this handled?

Reply:
> Upgraded to where? Does it need to be upgraded to the last patch version of the current minor, or to the minor you're wanting to upgrade cilium to? Or said differently, what do you envision the version compatibility of the SDP related to cilium to be? (This is assuming the SDP is versioned with cilium?)

i assume breaking changes would come in minor and not in patch. if current version is 1.17 and breaking changes coming in 1.18, then SDP has to be upgraded to 1.18 first and it should handle both 1.17 and 1.18 cilium bpf format and then upgrade cilium to 1.18.

Reply:
We tested sending the ip-identity updates to the SDP using the grpc method: https://github.com/vipul-21/cilium/commit/e70cbca5361b30efc7374b3c81a627972ae9b205#diff-b20ae8ce734da58871099f2c82dea704eb7f9ac2aea5c52541fb0572aa6962b1R386.
Observed a slight delay in the IP cache update listener. Consequently, if DNS requests and rules are applied simultaneously, there may be a situation where the rules are sent without the identity-to-IP mapping for a newly added identity, as the IP cache listener requires some time to update. In this situation, SDP will fail to serve the request. This will only happen momentarily as once the ip cache update happens SDP will get that updated data too.
In case of BFP map implementation, we did a load testing and did not see this problem while serving 1000 requests